integrations/cdi/datasource-ucp/src/main/java/io/helidon/integrations/datasource/ucp/cdi/UCPBackedDataSourceExtension.java

```diff
@@ -119,42 +121,152 @@ protected final void addBean(final BeanConfigurator<DataSource> beanConfigurator
+                        // factories (DataSource implementations,
+                        // usually) that can be fully configured with
+                        // the Universal Connection Pool to Strings
+                        // and those Strings which can be converted by
+                        // the PoolDataSourceImpl#toBasicType(String,
+                        // String) method.
+                        connectionFactoryProperties.setProperty(propertyName, properties.getProperty(propertyName));
+                    }
+                }
+            }
+            final Object serviceName = connectionFactoryProperties.remove("serviceName");
+            final Object pdbRoles = connectionFactoryProperties.remove("pdbRoles");
+            if (!connectionFactoryProperties.stringPropertyNames().isEmpty()) {
+                // We found some String-typed properties that are
+                // destined for the underlying connection factory to
+                // hopefully fully configure it.  Apply them here.
+                returnValue.setConnectionFactoryProperties(connectionFactoryProperties);
+            }
+            // Set the PoolDataSource's serviceName property so that
+            // it appears to the PoolDataSource to have been set via
+            // the undocumented XML configuration that the
+            // PoolDataSource can apparently be configured with in
+            // certain (irrelevant for Helidon) application server
+            // cases.
+            if (serviceName instanceof String) {
+                try {
+                    Method m = returnValue.getClass().getDeclaredMethod("setServiceName", String.class);
+                    if (m.trySetAccessible()) {
+                        m.invoke(returnValue, serviceName);
+                    }
+                } catch (final NoSuchMethodException ignoreOnPurpose) {
+
                 }
             }
-            // Permit further customization before the bean is actually created
-            instance.select(new TypeLiteral<Event<PoolDataSource>>() {}, dataSourceName).get().fire(returnValue);
+            // Set the PoolDataSource's pdbRoles property so that it
+            // appears to the PoolDataSource to have been set via the
+            // undocumented XML configuration that the PoolDataSource
+            // can apparently be configured with in certain
+            // (irrelevant for Helidon) application server cases.
+            if (pdbRoles instanceof Properties) {
+                try {
+                    Method m = returnValue.getClass().getDeclaredMethod("setPdbRoles", Properties.class);
+                    if (m.trySetAccessible()) {
+                        m.invoke(returnValue, pdbRoles);
+                    }
+                } catch (final NoSuchMethodException ignoreOnPurpose) {
+
+                }
+            }
+        }
+        final Instance<SSLContext> sslContextInstance = instance.select(SSLContext.class);
```

Comment:
Can we scope/restrict this to somehow related to datasource being produced here?

Reply:
Yeah, that's a good idea.